elbepack/initvmaction.py

```diff
@@ -173,7 +225,59 @@ def _attach_disk_fds(self):
                 os.open(source.attrib['file'], flags),
             ])
 
+    def _run_qemu_vm(self, initvmdir):
+        """
+        Start the initvm in QEMU mode.
+
+        This method starts the initvm in QEMU mode if not already another
+        initvm is running.
+        """
+
+        # for QEMU mode, the user needs to provide the path to the initvm directory.
+        if initvmdir == None:
+            if os.path.isdir('./initvm'):
+                # fallback to default path
+                print('Using default initvm directory "./initvm".')
+                initvmdir = './initvm'
+            else:
+                # no vm found - exit with error
+                print('No initvm found!')    
+                sys.exit(207)
+        
+        # test if there is already a process bound to the expected port
+        if is_qemu_running():
+            if os.path.exists(os.path.join(initvmdir, 'qemu-monitor-socket')):
+                # if the unix socket exists, assume this VM is bound ot the soap port
+                print('This initvm is already running.')
+            else:
+                # no unix socket file, assume another VM is bound to the soap port
+                print('There is already another running initvm.\nPlease stop this VM first.')
+                sys.exit(211)
+        else:
+            # try to start the QEMU VM for the given directory
+            system(f'cd {initvmdir} && make run_qemu &')
+
+            # wait for 10s that a process binds to the port
+            print('')
+            for i in range(10):
+                time.sleep(1)
+                if not is_qemu_running():
+                    print('*', end='', flush=True)
+                else:
+                    break
+            print('')
+
+            # test if we can communicate with the soap service
+            test_soap_communication(1)
+            print('initvm started successfully')
+
     def execute(self, _initvmdir, _opt, _args):
+        # handle QEMU mode
+        if _opt.qemu_mode:
+            self._run_qemu_vm(_initvmdir)
+            # action done for QEMU mode
+            return
```

Comment:
please also move the libvirt logic into its own function.

So we have:

```py
def execute(self, ...):
    if opt.qemu_mode:
        self._run_qemu(...)
    else:
        self._run_libvirt(...)
```

The same structure should also be used for other similar methods.

Reply:
Done